## Supplementary information {.page_break_before}

### Missense mutations in *Setmar* are unlikely to contribute to epistasis in the BXDs

Unlike *Ogg1*, *Setmar* does not participate directly in base-excision repair, though its primate ortholog plays an indirect role in the repair of double-stranded DNA breaks via non-homologous end joining (NHEJ). 
In anthropoid primates, *SETMAR* encodes a fusion of two functional domains: a SET domain-containing histone methyltransferase and a transposase domain from the *Mariner* family (MAR) [@PMID:16672366]; the mouse *Setmar* ortholog only encodes the histone methyltransferase domain.
In human cell lines, *SETMAR* localizes to induced double-strand breaks (DSBs) and dimethylates nearby H3K36, which promotes the recruitment of DNA repair components involved in NHEJ to the DSB [@PMID:21187428].
There is also evidence that overexpression of *SETMAR* (also known as *Metnase*) improves the efficiency of NHEJ [@PMID:16332963] and leads to increased cell survival following exposure to ionizing radiation [@PMID:16332963].
Point mutations in either the SET or MAR domains significantly reduced the ability of *SETMAR* to promote non-homologous end joining and DNA repair [@PMID:16332963;@PMID:24573677;@PMID:21491884], suggesting that both domains are needed for its role in DNA repair.
Another study found that overexpression of the isolated SET and MAR domains, but not of wild-type *SETMAR*, had a modest effect on NHEJ repair; overexpression of the SET domain slightly *decreased* NHEJ repair of a linearized plasmid in human cells, while overexpression of the *Mariner*-derived domain increased NHEJ relative to controls [@PMID:31238295].

Taken together, these results suggest that both the SET and transposase domains of primate *SETMAR* are important for *SETMAR*-mediated DNA repair. 
The p.Leu103Phe missense mutation that differentiates C57BL/6J and DBA/2J (Table @tbl:nonsyn-diffs) resides within the *Setmar* pre-SET domain and occurs at an amino acid residue that is predicted to be deleterious by SIFT [@PMID:12824425].
However, since the mouse *Setmar* ortholog lacks the *Mariner*-derived domain, we believe that the the p.Leu103Phe or p.Ser273Arg missense mutations are unlikely to affect C>A mutation rates in the BXDs. 
Moreover, we believe that the documented mutator phenotypes associated with *Ogg1*, as well as that gene's known role in base-excision repair, make it  more likely candidate to underlie the epistatic interaction with *Mutyh* we observed in this study.

### Supplementary Figures

![
**Simulations to assess the power of the aggregate mutation spectrum distance method.** 
In each of 50 trials, we simulated genotypes at 1,000 biallelic loci on a toy population of either 50 or 100 haplotypes as follows.
At every locus on every haplotype, we drew a single floating point value from a uniform distribution $[0, 1)$.
If that value was less than or equal to 0.5, we set the allele to be "A"; otherwise, we set the allele to be "B".
In each trial, we also simulated *de novo* germline mutations on the population of haplotypes, such that at a single locus $g_i$, we augmented the mutation rate of a particular $k$-mer by the specified effect size (an effect size of 1.5 indicates a 50% increase in the mutation rate) on haplotypes carrying "A" alleles.
We then applied the aggregate mutation spectrum distance method to these simulated data and asked if the adjusted cosine distance at locus $g_i$ was greater than expected by chance. 
Given a specific combination of parameters, the y-axis denotes the fraction of 50 trials in which the simulated mutator allele could be detected at a significance threshold of p = 0.05.
Shaded areas indicate the 95% bootstrap confidence interval surrounding that estimate.
](images/fig-power-simulations.png){#fig:simulations tag="1-figure supplement 1" width=7.5in} 

![
**Comparing power between the aggregate mutation spectrum distance method and QTL mapping.** 
In each of 50 trials, we simulated genotypes at 1,000 biallelic loci on a toy population of 100 haplotypes as follows.
At every locus on every haplotype, we drew a single floating point value from a uniform distribution $[0, 1)$.
If that value was less than or equal to 0.5, we set the allele to be "A"; otherwise, we set the allele to be "B".
In each trial, we also simulated *de novo* germline mutations on the population of haplotypes, such that at a single locus $g_i$, we augmented the rate of the specified mutation type by the specified effect size (an effect size of 1.5 indicates a 50% increase in the mutation rate) on haplotypes carrying "A" alleles.
We then applied the aggregate mutation spectrum distance method to these simulated data and asked if the adjusted cosine distance at locus $g_i$ was greater than expected by chance. 
Similarly, in each trial, we used R/qtl2 to perform a genome scan for QTL and asked if the log-odds score at $g_i$ was greater than expected by chance.
Given a specific combination of parameters, the y-axis denotes the fraction of 50 trials in which the simulated mutator allele could be detected at a significance threshold of p = 0.05 (for AMSD) or at an alpha of $\frac{0.05}{7}$ (for QTL mapping).
Shaded areas indicate the 95% bootstrap confidence interval surrounding that estimate.
](images/fig-power-comparison.png){#fig:ihd_vs_qtl_power tag="1-figure supplement 2" width=7.5in} 

![
**Comparing power between the aggregate mutation spectrum distance method and QTL mapping with variable counts of simulated mutations.** 
In each of 50 trials, we simulated genotypes at 1,000 biallelic loci on a toy population of 50 or 100 haplotypes as follows.
At every locus on every haplotype, we drew a single floating point value from a uniform distribution $[0, 1)$.
If that value was less than or equal to 0.5, we set the allele to be "A"; otherwise, we set the allele to be "B".
In each trial, we also simulated *de novo* germline mutations on the population of haplotypes, such that at a single locus $g_i$, we augmented the rate of the specified mutation type by the specified effect size (an effect size of 1.5 indicates a 50% increase in the mutation rate) on haplotypes carrying "A" alleles.
To more closely approximate the BXD RILs, the mean number of simulated mutations on each haplotype was allowed to vary by a factor of 20 (see Materials and Methods for more details).
We then applied the aggregate mutation spectrum distance method to these simulated data and asked if the adjusted cosine distance at locus $g_i$ was greater than expected by chance. 
Similarly, in each trial, we used R/qtl2 to perform a genome scan for QTL and asked if the log-odds score at $g_i$ was greater than expected by chance.
Given a specific combination of parameters, the y-axis denotes the fraction of 50 trials in which the simulated mutator allele could be detected at a significance threshold of p = 0.05 (for AMSD) or at an alpha of $\frac{0.05}{7}$ (for QTL mapping).
Shaded areas indicate the 95% bootstrap confidence interval surrounding that estimate.
](images/fig-power-comparison-variable-counts.png){#fig:ihd_vs_qtl_power_variable_counts tag="1-figure supplement 3" width=7.5in} 

![
**Comparing power between the aggregate mutation spectrum distance method and QTL mapping with variable mutator allele frequencies.** 
In each of 50 trials, we simulated genotypes at 1,000 biallelic loci on a toy population of 100 haplotypes as follows.
At every locus on every haplotype, we drew a single floating point value from a uniform distribution $[0, 1)$.
If that value was less than or equal to 0.5, we set the allele to be "A"; otherwise, we set the allele to be "B".
To model the effects of mutator allele frequencies on AMSD and QTL power, we allowed the expected frequency of "A" alleles at the mutator allele marker to be either 0.1, 0.25, or 0.5 in these simulations.
In each trial, we also simulated *de novo* germline mutations on the population of haplotypes, such that at a single locus $g_i$, we augmented the rate of the specified mutation type by the specified effect size (an effect size of 1.5 indicates a 50% increase in the mutation rate) on haplotypes carrying "A" alleles.
We then applied the aggregate mutation spectrum distance method to these simulated data and asked if the adjusted cosine distance at locus $g_i$ was greater than expected by chance. 
Similarly, in each trial, we used R/qtl2 to perform a genome scan for QTL and asked if the log-odds score at $g_i$ was greater than expected by chance.
Given a specific combination of parameters, the y-axis denotes the fraction of 50 trials in which the simulated mutator allele could be detected at a significance threshold of p = 0.05 (for AMSD) or at an alpha of $\frac{0.05}{7}$ (for QTL mapping).
Shaded areas indicate the 95% bootstrap confidence interval surrounding that estimate.
](images/fig-power-comparison-variable-afs.png){#fig:ihd_vs_qtl_power_variable_afs tag="1-figure supplement 4" width=7.5in} 

![
**Quantitative trait locus scans for mutation spectrum phenotypes.**
Using the BXDs with D genotypes at `rs27509845` (the marker with the highest cosine distance on chromosome 4; n = 66 BXDs, 42,171 total mutations), we used R/qtl2 to perform QTL scans for the fractions of each 1-mer mutation type.
QTL scans also included a kinship matrix (that contained the pairwise genetic similarity between each pair of BXDs, calculated using the leave-one-chromosome-out method) as a random effect term using the `kinship` keyword argument in the `scan1` function.
Plots show the log-odds (LOD) score at every genotyped marker in blue; the dotted black line represents the genome-wide LOD significance threshold (established using 1,000 permutations at an alpha of $\frac{0.05}{7}$ to account for the fact that 7 separate association tests were performed.)
](images/fig-qtl-scans.png){#fig:qtl-scans tag="2-figure supplement 1" width=7.5in} 

![
**Mutation spectra comparison in BXD strains.**
Fractions of *de novo* germline mutations in BXDs with either *D* or *B* genotypes at markers `rs27509845` and `rs46276051`, stratified by mutation type.
](images/fig-spectra-comparison-bxd-all.png){#fig:spectra-comparison-all tag="3-figure supplement 1" width=7.5in}

![
**Mutation spectra comparison in Sanger Mouse Genomes Project strains.**
Fractions of *de novo* germline mutations in Sanger MGP strains with either *D* or *B* haplotypes at the chromosome 4 and chromosome 6 mutator loci, stratified by mutation type.
](images/fig-spectra-comparison-mgp.png){#fig:spectra-comparison-mgp tag="3â€”figure supplement 2" width=7.5in} 

![
**Frequency of nonsynonymous DNA repair mutations in wild mice.** Alternate allele frequencies of each nonsynonymous DNA repair mutation overlapping the chromosome 6 mutator locus were calculated in populations of wild-derived mice from Harr et al. [@PMID:27622383]. Numbers of mice in each subpopulation are shown in parentheses. *Mmc* (*Mus musculus castaneus*), *Mmd* (*Mus musculus domesticus*), *Mmm* (*Mus musculus musculus*), and *Ms* (*Mus spretus*). The *Mbd4* p.Asp129Asn mutation was not observed in any wild populations.
](images/fig-wild-afs.png){#fig:wild-afs tag="3-figure supplement 3" width=7.5in}


### Supplementary Tables

| Gene name | Tissue name | # BXDs with expression data |  Top significant marker | -log10(p) at top significant marker (GEMMA) | Additive effect of D allele on expression (GEMMA) |
| - | - | - | - | - | - |
| *Ogg1* | Kidney | 53 | `rsm10000004188` | 12.89 | -0.180 |
| *Ogg1* | Liver | 50 | `rsm10000004188` | 13.57 | -0.155 | 
| *Ogg1* | Spleen | 79 | `rsm10000003418` | 4.73 | -0.056 |
| *Ogg1* | Gastrointestinal | 46 | `rs4173870` | 5.43 | -0.048 |
| *Fancd2* | Gastrointestinal | 46 | `rsm10000004199` | 8.60 | 0.133 | 
| *Ogg1* | Hippocampus | 67 | `rsm10000004188` | 16.50 | -0.165 |
| *Rad18* | Hippocampus | 67 | `rsm10000003463` | 6.32 | 0.068 |
| *Setmar* | Hippocampus | 67 | `rs13478947` | 11.03 | 0.141 |
| *Mbd4* | Spleen | 79 | `rsm10000004199` | 6.05 | 0.071 | 

Table: Significant cis-eQTLs for DNA repair genes in various tissues identified using GeneNetwork. {#tbl:eqtl-results tag="supplement 1"}

<!-- | SV start | SV end |  SV type | Gene name(s) | Overlaps exon? |
| - | - | - | - | - |
| 112,629,618 | 112,636,619 | DEL | *Rad18* | No | 

Table: Large structural variants overlapping protein-coding genes in the mutator locus on chromosome 6. All coordinates are with respect to GRCm39/mm39. {#tbl:sv-overlap tag="supplement 2"} -->

